docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
+vec {}
+vec { "john@doe.com"; "john.doe@example.com" };
+....
+
+Subtypes::
+
+* Whenever `t` is a subtype of `t'`, then `vec t` is a subtype of `vec t'`.
+* `blob` is a subtype of `vec nat8`
+
+Supertypes::
+
+* Whenever `t` is a supertype of `t'`, then `vec t` is a supertype of `vec t'`.
+* `blob` is a supertype of `vec nat8`
+
+Corresponding Motoko type::
+
+`[T]`, where the Motoko type `T` corresponds to `t`
+
+Corresponding Rust type::
+
+`Vec<T>`, where the Rust type `T` corresponds to `t`
+
+Corresponding JavaScript values::
+
+TODO:
+
+[#type-opt]
+=== Type opt t
+
+The `opt t` type contains of the values of type `t`, plus the special `null` value. It is used to express that some value is optional, i.e. it may be present, or there may be no value.
+
+The `opt` type can be nested (e.g. `opt opt text`), and the values `null` and `opt null` are distinct values.
+
+The `opt` type plays a crucial role in the evolution of Candid interfaces, and has atypical subtyping rules; see below.
+
+Type syntax::
+
+`opt bool`, `opt nat8`, `opt opt text`,…
+
+Textual syntax::
++
+....
+null
+opt true
+opt 8
+opt null
+opt opt "test"
+....
+
+Subtypes::
+
+The canonical rules for subtyping with `opt` are:
++
+* Whenever `t` is a subtype of `t'`, then `opt t` is a subtype of `opt t'`.
+* `null` is a subtype of `opt t'`.
+* `t` is a subtype of `opt t` (unless `t` itself is `null`, `opt …` or `reserved`)
++
+In addition, for technical reasons related to upgrading and higher-order services, _every_ type is a subtype of `opt t`. Users are advised, however, to not directly make use of that rule.
+
+Supertypes::
+
+* Whenever `t` is a supertype of `t'`, then `opt t` is a supertype of `opt t'`.
+
+Corresponding Motoko type::
+
+`?T`, where the Motoko type `T` corresponds to `t`
+
+Corresponding Rust type::
+
+`Option<T>`, where the Rust type `T` corresponds to `t`
+
+Corresponding JavaScript values::
+
+TODO:
+
+[#type-record]
+=== Type record { n : t, … }
+
+A `record` type is a collection of labeled values. For example,
+....
+type address = record {
+  street : text;
+  city : text;
+  zip_code : nat;
+  country : text;
+};
+....
+gives the name `address` to the type of records that have textual fields `street`, `city` and `country` and a numerical field `zip_code`. The order of fields in the record type declaration does not matter.
+
+The label of a record field can also be a 32-bit natural number, as in
+....
+type address2 = record {
+  288167939 : text;
+  1103114667 : text;
+  220614283 : nat;
+  492419670 : text;
+};
+....
+In fact, textual labels are treated as their _field hash_, and incidentially, `address` and `address2` are – to Candid – the same types.
+
+If you omit the label, Candid numbers automatically assigns sequentially increasing labels. This leads to the following short syntax, which is typically used to represent pairs and tuples. The type `record { text; text; opt bool }` is equivalent to `record { 0 : text;  1: text;  2: opt bool }`
```

Comment:
```suggestion
If you omit the label, Candid automatically assigns sequentially increasing labels. This leads to the following short syntax, which is typically used to represent pairs and tuples. The type `record { text; text; opt bool }` is equivalent to `record { 0 : text;  1: text;  2: opt bool }`
```
Does this only kick in if you omit all labels? or does it start numbering from the previous label? Does that even make sense for record values where you'd expect to be able to provide labeled values in any order?

Reply:
It starts numbering from the prevoius label, but I made the opinionated choice to _not_ explain that in the user’s manual (I don't really see a good use case for it). The specification has all the answers…